docs/design/chart_format.md

```diff
@@ -0,0 +1,664 @@
+
+
+Charts must not assume that they can perform file system operations to load another chart or supporting resources directly via the filesystem, nor should they store any operational files outside of the chart directory. It should be possible to archive the chart directory (e.g. `tar -cf ROOT.tar ROOT/`) and not lose any necessary information.
+
+### Generators and Templates
+The contents inside of the `templates/` directory may contain directives that can be processed by an external processor. Such a processor may read these files and generate new content.
+
+
+A generator reads through the files in `templates/` and looks for special file headers that indicate what generation tool needs to be run (if any) on a file. One such implementation of a generation tool is a template renderer. In any one pass through the `templates/` directory, more than one generation tool may be invoked.
+
+
+#### The Generator Header
+A file that requires a generation tool must declare this on the first line of the file using one of the following syntaxes:
+
+```
+//helm:generate TOOL [ARGS]\n
+#helm:generate TOOL [ARGS]\n
+/*helm:generate TOOL [ARGS]*/
+```
+
+`TOOL` corresponds to a registered tool name (see [this proposal](https://github.com/kubernetes/kubernetes/pull/18215#issuecomment-168885511)). `[ARGS]` are zero or more space-separated strings that will be passed unmodified to the tool. The first two formats require a terminating line feed. The third format requires a terminating `*/` sequence.
+
+
+**Example:**
+
+```
+#helm:generate dm_template
+{% set PROPERTIES = properties or {} %}
+{% set WORKERS = PROPERTIES['workers'] or 2 %}
+resources:
+- name: spark-master
+ type: ReplicationController
+ properties:
+   kind: ReplicationController
+   apiVersion: v1
+   metadata:
+     name: spark-master-controller
+   spec:
+…
+```
+
+
+In the example above, the `#helm:generate dm_template` indicates that the remainder of the file contents should be processed using whatever handler is registered for handling `dm_template`.
+
+
+Generator tools may take arguments. No format is specified for arguments. For example, one might use UNIX-like flags.
+
+
+```
+#helm:generate dm_template --ignore-timezone -s
+{% set PROPERTIES = properties or {} %}
+{% set WORKERS = PROPERTIES['workers'] or 2 %}
+resources:
+- name: spark-master
+ type: ReplicationController
+ properties:
+   kind: ReplicationController
+   apiVersion: v1
+   metadata:
+     name: spark-master-controller
+   spec:
+…
+```
+
+Any arguments are passed into the generator unaltered. Generators may ignore arguments that they do not support or understand.
+
+
+A generator tool _must_ not overwrite any files (including itself) present in the chart’s `templates/` directory. It must not alter the `README.md`, the `LICENSE`, the `Chart.yaml`, or the contents of the `hooks/` directory. These restrictions maintain the reproducibility of the build over subsequent generator runs.
```

Comment:
@jackgr @sparkprime Considering this is a chart format definition doc, do you think it would be appropriate to go into the details of the generator in a separate design doc or do we want it here?